components/style/values/computed/length.rs

```diff
@@ -76,6 +76,17 @@ impl CalcLengthOrPercentage {
     pub fn percentage(&self) -> CSSFloat {
         self.percentage.unwrap_or(0.)
     }
+
+    #[allow(missing_docs)]
+    pub fn calc(&self, length: Option<Au>) -> Au {
```

Comment:
This need some changes. We want to judge the situation that the `percentage` field is `Some` and container length is `None` that would resolve to a auto length no matter what the `length` field is. `MaybeAuto` type is only present in `layout` you can either return an `Option<Au>` here and add a function to convert it to `MaybeAuto` in `layout/model.rs` since they are isomorphic, or add a Trait returning the `MaybeAuto` type in layout and implement it for `CalcLengthOrPercentage`.

BTW, please select some more semantic names :)